File: week_1/day-5-string-hashing/first_non_repeating_char_follow_up_questions/first_non_repeating_optimized.py
# Program: first_non_repeating_optimized.py
def firstUniqCharOptimized(s):
    count = [0] * 26

    for ch in s:
        count[ord(ch) - ord('a')] += 1

    for i, ch in enumerate(s):
        if count[ord(ch) - ord('a')] == 1:
            return i

    return -1

print(firstUniqCharOptimized("leetcode"))  # Output: 0

File: week_1/day-5-string-hashing/first_non_repeating_char_follow_up_questions/firstNonRepeatingOptimized.js
// Program: firstNonRepeatingOptimized.js
function firstUniqCharOptimized(s) {
    const count = new Array(26).fill(0);

    for (let ch of s) {
        count[ch.charCodeAt(0) - 97]++;
    }

    for (let i = 0; i < s.length; i++) {
        if (count[s.charCodeAt(i) - 97] === 1) return i;
    }
    return -1;
}

console.log(firstUniqCharOptimized("leetcode")); // 0
